### ISSUE TITLE
AttributeError: 'module' object has no attribute 'PKCS5_SALT_LEN'

### ISSUE THREAD

**Issue by destijl** [M2Crypto/M2Crypto#24]
Looks like people have filed issues in a bunch of other places, but not with the actual project.  See:

https://github.com/georgebrock/1pass/issues/21
https://bugs.gentoo.org/show_bug.cgi?id=538920

The error is:

$ python -c "import M2Crypto"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "build/bdist.macosx-10.8-x86_64/egg/M2Crypto/__init__.py", line 24, in <module>
  File "build/bdist.macosx-10.8-x86_64/egg/M2Crypto/ASN1.py", line 12, in <module>
  File "build/bdist.macosx-10.8-x86_64/egg/M2Crypto/BIO.py", line 221, in <module>
  File "build/bdist.macosx-10.8-x86_64/egg/M2Crypto/BIO.py", line 227, in CipherStream
AttributeError: 'module' object has no attribute 'PKCS5_SALT_LEN'

$ python
Python 2.7.9 (default, Jan 29 2015, 06:28:46) 
[GCC 4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.40)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import m2
>>> m2.PKCS5_SALT_LEN
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'PKCS5_SALT_LEN'

I verified the 'fix' listed in the Gentoo bug, i.e. using swig 3.0.4 works fine, and 3.0.5 is broken.  My environment was OS X 10.8.5 with M2Crypto-0.22.3.

**Comment by danmcd** [comment]
I'm being bitten by this as well.  Is the workaround to keep SWIG at 3.0.4?  (I upgraded swig from 2.0.12 right to 3.0.5, and it appears that was a Bad Idea (TM).)

**Comment by guilbep** [comment]
@danmcd I switched back to swig 3.0.4 ; rebuild m2crypto and it worked...

**Comment by microcai** [comment]
keep SWIG at 3.0.4 clearly is not a FIX, who is stupid enough to mark it fixed ? What the fuck are  you doing?

**Comment by bsmartt13** [comment]
Thanks for the workaround it works for me!

**Comment by pyeekwr** [comment]
I'm running into this problem as well with swig 3.0.5

**Comment by isokrates** [comment]
Hey!

Would someone be so kind as to treat me like a complete idiot and explain precisely the steps one might take to overcome this error on OS X 10.10?

I got swig 3.0.5 and M2Crypto 0.22.3 installed (after what seemed like a lot of trial and error!), and am running into the same problem ('module' object has no attribute 'PKCS5_SALT_LEN').

I see from this comment thread that installing swig 3.0.4 in place seems to fix it, but can't work out how to do so - `brew install swig` seems to have installed 3.0.2, but swig -version shows 3.0.5 (probably because I downloaded this version from swig.org and installed it first - running `brew install swig` again says that 3.0.2 is installed but not linked).

In short, what's the safest way to remove swig 3.0.5 (and probably the 3.0.2 I installed with Homebrew), and install 3.0.4 in its place, and should I also uninstall then reinstall M2Crypto (and if so, what's the safest way to do that)?

Apologies for what is probably a very basic question, but I'm somewhat new to this!

Cheers :)

**Comment by tricoder42** [comment]
@isokrates This worked for me about an hour ago...

```
# Cleanup
pip uninstall m2crypto
brew uninstall swig --force

# Install swig 3.0.4 and compile m2crypto
brew update
brew install homebrew/versions/swig304
brew link homebrew/versions/swig304
pip install m2crypto
```

**Comment by isokrates** [comment]
Amazing - works for me too.

Thank you so much!! :)

**Comment by y2h4ck** [comment]
@tricoder42 
Thanks for your tip!
Worked as a charm! :-)

**Comment by ryanneufeld** [comment]
What is required to fix M2Crypto to resolve this bug?

**Comment by y2h4ck** [comment]
Man this is a very annoying problem. I solve this way
first clean up

$pip uninstall m2crypto

then uninstall swig

after ... reinstall and relink swig 3.0.4
and after reinstall m2crypto.


this will solve.





2015-03-23 19:31 GMT-03:00 Ryan Neufeld <notifications@github.com>:

> What is required to fix M2Crypto to resolve this bug?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/M2Crypto/M2Crypto/issues/24#issuecomment-85232112>.
>



-- 
att


<http://about.me/anderson.tamborim>

Anderson Tamborim / Social Engineer & Penetration Testery2h4ck@gmail.com

http://about.me/anderson.tamborim

[image: Facebook]
<https://www.facebook.com/profile.php?id=100001040414830>[image:
Twitter]  <https://twitter.com/y2h4ck>[image: Google Plus]
<https://plus.google.com/u/0/101478142294165683789>[image: Linkedin]
<http://br.linkedin.com/in/atamborim>[image: skype]

**Comment by ryanneufeld** [comment]
I was able to do this yes, but it's a pretty poor workaround to have to
downgrade one library to support another.

--Ryan Neufeld BSc CIS
ryan@neufeldmail.com

**Comment by y2h4ck** [comment]
Yeah it sux, but this is the way I found to make work on Mac OSX. If you
find another way please tell me :)
best regards!

2015-03-25 14:01 GMT-03:00 Ryan Neufeld <notifications@github.com>:

> I was able to do this yes, but it's a pretty poor workaround to have to
> downgrade one library to support another.
>
> --Ryan Neufeld BSc CIS
> ryan@neufeldmail.com
>
> On Mon, Mar 23, 2015 at 5:45 PM, y2h4ck <notifications@github.com> wrote:
>
> > Man this is a very annoying problem. I solve this way
> > first clean up
> >
> > $pip uninstall m2crypto
> >
> > then uninstall swig
> >
> > after ... reinstall and relink swig 3.0.4
> > and after reinstall m2crypto.
> >
> >
> > this will solve.
> >
> >
> >
> >
> >
> > 2015-03-23 19:31 GMT-03:00 Ryan Neufeld <notifications@github.com>:
> >
> > > What is required to fix M2Crypto to resolve this bug?
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/M2Crypto/M2Crypto/issues/24#issuecomment-85232112
> >.
> > >
> >
> >
> >
> > --
> > att
> >
> >
> > <http://about.me/anderson.tamborim>
> >
> > Anderson Tamborim / Social Engineer & Penetration Testery2h4ck@gmail.com
> >
> > http://about.me/anderson.tamborim
> >
> > [image: Facebook]
> > <https://www.facebook.com/profile.php?id=100001040414830>[image:
> > Twitter] <https://twitter.com/y2h4ck>[image: Google Plus]
> > <https://plus.google.com/u/0/101478142294165683789>[image: Linkedin]
> > <http://br.linkedin.com/in/atamborim>[image: skype]
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/M2Crypto/M2Crypto/issues/24#issuecomment-85264637>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/M2Crypto/M2Crypto/issues/24#issuecomment-86121860>.
>



-- 
att


<http://about.me/anderson.tamborim>

Anderson Tamborim / Social Engineer & Penetration Testery2h4ck@gmail.com

http://about.me/anderson.tamborim

[image: Facebook]
<https://www.facebook.com/profile.php?id=100001040414830>[image:
Twitter]  <https://twitter.com/y2h4ck>[image: Google Plus]
<https://plus.google.com/u/0/101478142294165683789>[image: Linkedin]
<http://br.linkedin.com/in/atamborim>[image: skype]

**Comment by ryanneufeld** [comment]
I'd like the maintainers to fix it.

--Ryan Neufeld BSc CIS
ryan@neufeldmail.com

**Comment by mcepl** [comment]
This is apparently https://github.com/swig/swig/issues/344 ... I will take a look at it.

**Comment by someshchadda** [comment]
Hi,
I am new to salt, I am also facing the same issue.
I have tried the steps mentioned by tricoder42, but I am still facing the same issue.

**Comment by bbhoss** [comment]
Thanks for the workaround. Installing saltstack --HEAD via homebrew, it seems your need to `rm -rf /usr/local/Cellar/saltstack/HEAD/libexec/lib/python2.7/site-packages/M2Crypto`

**Comment by tingtingtwice** [comment]
This is so wonderful. Thank you tricoder42 !

**Comment by joejulian** [comment]
Ok, I was able to get it to load after compiling with 3.0.5, but I haven't figured out how to get the generated SWIG/m2crypto.py to install with distutils yet..

https://github.com/M2Crypto/M2Crypto/compare/master...joejulian:use_swig_3_0_5

**Comment by gregwym** [comment]
Make sure you clean the pip cache folder before rebuild M2Crypto after reinstall swig 3.0.4.

**Comment by kittozheng** [comment]
Thanks @tricoder42 ! It works! : )
